### PR TITLE
Adding output to JUnit.xml with JUnit schema files.

### PR DESCRIPTION
**Pull request by QGhd**
I have attached two changed source files (replacing src/testdrive.f90 and test/main.f90) as well as the resulting JUnit.xml file that is written along with the screen output. This xml is doing the job for us in an Azure DevOps pipeline based on the data that is also written on the screen.
The main.f90 uses 4 additional external subroutines from testdrive.f90. Two to open and close the Junit.xml and two to open and close the xml tag. The rest of the xml output is covered internally in testdrive.f90.
However, I assume that you would prefer to avoid these additional functions, but I choose a simple approach with minimal code changes, instead of a major rework of the module structure.

In addition, the JUnit.xml file references to an xml schema (JUnit.xsd from https://github.com/windyroad/JUnit-Schema). In theory the JUnit.xml should be conform to this schema, but it is not for several reasons:

There are missing (required) xml tag attributes, like number of all/failed/succeeded/skipped tests per testsuite that are as summarized data not available at the time, when the according opening xml tag is written.
There are missing xml tag attributes, like execution time, that are not stored in the code (as far as I could see).
There is no corresponding xml tag in the schema file for the additional stdout/stderr output per testcase. Therefore, I added according xml tags () per testcase that are - following the xsd - only allowed once per testsuite.
Fortunately, Azure DevOps does not care about the discrepancy, so we are fine with this setup (but it is not a generic solution).
I have also attached another xml schema file (jenkins-junit.xsd from https://github.com/junit-team/junit5 resp. https://junit.org/junit5/). To be honest I am not sure, what should be regarded as the standard for a JUnit.xml. In fact, I do not care much, since Azure DevOps test pipeline works fine. Others, however, may have a more specific opinion on the subject.

Let me know, if you want to discuss my changes or whether I can support you.
Thank you for considering my changes.

### CODE REVIEW

**Comment by jvdp1** [comment]
Interesting. the JUnit.xml file could be acutally also used in the CI/CD of Gitlab. The current generated xml file is already recognized by Gitlab (still a few issues with the name of the suite not recognized by Gitlab). Execution time would be useful too.

File: src/testdrive.F90 [hunk 1]
```diff
@@ -306,9 +312,150 @@ end subroutine collect_interface
...
 
-
-contains
-
+  contains
+  
+  !> Open JUnit.xml file for CLI output of test results.
+  subroutine junitxml_open_file()
+  
+    implicit none
```
Comment (jvdp1) [inline src/testdrive.F90:320]: These `implicit none` are not needed, as it is inside a module that already states it.

File: src/testdrive.F90 [hunk 2]
```diff
@@ -306,9 +312,150 @@ end subroutine collect_interface
...
+    if (unit_junitxml /= -1) then
+      write(unit_junitxml,'(a)') '<?xml version="1.0" encoding="UTF-8"?>'
+      write(unit_junitxml,'(a)') &
+        & '<testsuites' // &
+        & ' xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"' // & 
+        & ' xsi:noNamespaceSchemaLocation="JUnit.xsd"' // & 
+        & '>'
+    else
+      write(error_unit, '(a)') "# Error: Could not open JUnit.xml for writing! Program stops."
+      stop 1
```
Comment (jvdp1) [inline src/testdrive.F90:333]: Maybe:
```suggestion
      error stop 1
```

File: src/testdrive.F90 [hunk 3]
```diff
@@ -396,6 +543,9 @@ recursive subroutine run_unittest(test, unit, stat)
 
     type(error_type), allocatable :: error
     character(len=:), allocatable :: message
+    character(len=20) :: res
```
Comment (jvdp1) [inline src/testdrive.F90:546]:
```suggestion
    character(len=:), allocatable :: res
```

File: src/testdrive.F90 (hunk 1 above)
Comment (QGhd) [reply src/testdrive.F90:320]: Accepted.

File: src/testdrive.F90 (hunk 3 above)
Comment (QGhd) [reply src/testdrive.F90:546]: Accepted.

File: src/testdrive.F90 [hunk 4]
```diff
@@ -405,6 +555,34 @@ recursive subroutine run_unittest(test, unit, stat)
...
+        else
+          call junitxml_write_testcase_single_tag(test%name)
+        endif
+      case ('SKIPPED')
+        res = res
+      case default
+        write(unit, '(a)') "Error: Unknown test result '" // res // "' in test '" // test%name // "'!"
+        stop 2
+    end select
+
```
[the remark this answers is not in the document]
Comment (QGhd) [reply src/testdrive.F90:585]: I am not sure, what is the best solution. I am not experienced with OpenMP.
Wrapping the calls in
`!$omp critical(testdrive_testsuite)
call junitxml_...()
!$omp end critical(testdrive_testsuite)` seems too be quite coarse.
I could add the `!$omp critical ...` around any `write` statement inside the calls.
But does this resolve the OpenMP parallel issue?

File: .gitignore [hunk 5]
```diff
@@ -35,3 +35,5 @@
 /build*/
 /_*/
 /docs*/
+/.vs
+/out/build/x64-Debug
```
Comment (awvwgk) [inline .gitignore:39]: These changes are unrelated to this PR.

File: src/JUnit.xml
Comment (awvwgk) [inline src/JUnit.xml:1]: Should this file be checked in?

File: src/JUnit.xsd
Comment (awvwgk) [inline src/JUnit.xsd:1]: Same for this file, should this be checked in?

File: src/jenkins-junit.xsd
Comment (awvwgk) [inline src/jenkins-junit.xsd:1]: This also seems to be checked in by accident

**Review by awvwgk — CHANGES_REQUESTED** [review]
Thanks, this looks pretty straight forward to add and I would be happy to accept this in test-drive. One general comment, the generation of the JUnit.xml is incremental, which means if the test run is aborted the partly generated JUnit.xml file is always invalid. Could we have a process which guarantees that the JUnit.xml file can only be written completely?

File: .gitignore (hunk 5 above)
Comment (QGhd) [reply .gitignore:39]: Will change.

File: src/JUnit.xml
Comment (QGhd) [reply src/JUnit.xml:1]: I have added for you (resp. the reviewers). It documents the expected output. Could be used for your documentation. Could also be used, if you would like to implement a test. But it is not relevant for the build and can be deleted, if you want.

File: src/jenkins-junit.xsd
Comment (QGhd) [reply src/jenkins-junit.xsd:1]: Also for documentation resp. for you. This schema file might be an alternative for someone, how prefers this schema file for Jenkins. I cannot test it. If you prefer, I can delete any of these files from the PR.

File: .gitignore (hunk 5 above)
Comment (QGhd) [reply .gitignore]: Reverted.

**Comment by QGhd** [comment]
> Thanks, this looks pretty straight forward to add and I would be happy to accept this in test-drive. One general comment, the generation of the JUnit.xml is incremental, which means if the test run is aborted the partly generated JUnit.xml file is always invalid. Could we have a process which guarantees that the JUnit.xml file can only be written completely?

Good point. Originally, I tried to use a FINAL statement for the module to make sure that the final closing tag is written and the xml file is closed, even in the error case. However, I failed. My final subroutine was never executed.
In any case the nested structure of the xml file requires either to track the current depth, when ABORT, ERROR, ... occurs or to use one FINAL statement per interface of the module.
What do you think?

File: src/testdrive.F90 [hunk 6]
```diff
@@ -405,6 +539,34 @@ recursive subroutine run_unittest(test, unit, stat)
...
+          call junitxml_write_testcase_closing_tag(stdout)
+        else
+          call junitxml_write_testcase_single_tag(test%name)
+        endif
+      case ('SKIPPED')
+        res = res
+      case default
+        write(unit, '(a)') "Error: Unknown test result '" // res // "' in test '" // test%name // "'!"
+        error stop 2
+    end select
```
Comment (jvdp1) [inline src/testdrive.F90:568]: Maybe a suggestion to avoid the issue with I/O in the parallel loop:

Transform the subroutines `junit_write_testcase_*` in functions returning `charachter` variables, concatanate them in one single `character` variable, and write it in a critical section. This will also avoid the incremental issue raised by @awvwgk 

this could be additionally extended to the whole JUnit.xml file, that is the content of the JUnit.xml file will be first saved in a `character` variable, and be written to the file only if completed.

File: src/testdrive.F90 [hunk 7]
```diff
@@ -405,6 +539,34 @@ recursive subroutine run_unittest(test, unit, stat)
...
+    
+    stdout = ''
+    if (allocated(error)) then
+      stdout = trim(error%message)
+    endif
+    
+    s = index(message, '[')
+    e = index(message, ']')
+    res = message(s+1:e-1)
+    select case (res)
```
Comment (jvdp1) [inline src/testdrive.F90:551]: Does this mean that this `select case` also happen, even if the user does not want JUnit.xml file?

Comment (QGhd) [reply src/testdrive.F90:551]: At the moment the JUnit.xml is written always.

File: src/testdrive.F90 (hunk 1 above)
Comment (QGhd) [reply src/testdrive.F90]: @jvdp1: I am not sure about the code of conduct.
Do you resolve this conversation? Or is it okay, if I do it?

File: src/JUnit.xml
Comment (QGhd) [reply src/JUnit.xml:1]: @awvwgk: Do you have preference, whether to delete this file or not?
It is your project, so I will follow your choice.

File: src/JUnit.xsd
Comment (QGhd) [reply src/JUnit.xsd:1]: @awvwgk: Do you have preference, whether to delete this file or not?
It is your project, so I will follow your choice.

File: src/jenkins-junit.xsd
Comment (QGhd) [reply src/jenkins-junit.xsd:1]: @awvwgk: Do you have preference, whether to delete this file or not?
It is your project, so I will follow your choice.

File: src/testdrive.F90 (hunk 2 above)
Comment (QGhd) [reply src/testdrive.F90]: @jvdp1 : Can I resolve the issue?

File: src/testdrive.F90 (hunk 1 above)
Comment (jvdp1) [reply src/testdrive.F90]: Either way is fine.

File: src/testdrive.F90 (hunk 7 above)
Comment (jvdp1) [reply src/testdrive.F90:551]: Up to @awvwgk to decide the final behaviour, but my preference would be htat JUnit.xml is written only on request of the user.

File: src/testdrive.F90 [hunk 8]
```diff
@@ -306,9 +312,134 @@ end subroutine collect_interface
...
+  
+  !> Write opening tag for testsuite with name to JUnit.xml.
+  subroutine junitxml_write_testsuite_opening_tag(testsuite_name, id)
+  
+    character(len=*), intent(in) :: testsuite_name
+    integer, intent(in) :: id
+    
+    character(len=80) :: hostname
+    character (len=8)  cdate
+    character (len=10) ctime
```
Comment (jvdp1) [inline src/testdrive.F90:352]:
```suggestion
    character(len=8)  :: cdate
    character(len=10) :: ctime
```

File: src/testdrive.F90 [hunk 9]
```diff
@@ -405,6 +539,34 @@ recursive subroutine run_unittest(test, unit, stat)
...
+      case ('FAILED', 'UNEXPECTED PASS')
+        call junitxml_write_testcase_opening_tag(test%name)
+        call junitxml_write_testcase_failure(res)
+        call junitxml_write_testcase_closing_tag(stdout)
+      case ('EXPECTED FAIL', 'PASSED')
+        if (len_trim(stdout) > 0) then  
+          call junitxml_write_testcase_opening_tag(test%name)
+          call junitxml_write_testcase_closing_tag(stdout)
+        else
+          call junitxml_write_testcase_single_tag(test%name)
```
Comment (jvdp1) [inline src/testdrive.F90:561]: All these subroutines could be combined into a single one:
```fortran
call  junitxml_write_testcase(test%name, res, stdout)
```
with `res` and `stdout` being optional.

it could even be a function to be used as:
```
!$omp critical
select case(res)
 case(...)
  write(unit_junitxm, '(a)') junitxml_write_testcase(test%name, res, stdout)
 case(...)
end select
!$end omp critical
```
Such an approach (with a function returning a character string, instead of writing it directly to a file) could solve the issue raised by @awvwgk

File: src/testdrive.F90 (hunk 6 above)
Comment (QGhd) [reply src/testdrive.F90:568]: Sorry, for not responding earlier. I was on holiday the last weeks.
I looked into the proposal to return a string instead of write it to the file. But I am not sure, whether it is a pragmatic way forward. I see two options:

1. In FORTRAN 2003+ one can use dynamic allocatable strings on every level of the code (see https://stackoverflow.com/questions/14886787/fortran-character-input-at-undefined-length, https://stackoverflow.com/questions/31084087/using-a-deferred-length-character-string-to-read-user-input).
2. In older FORTRAN (< 2003) the string size needs to be managed on gloabl scope to grow with the amount of test output, since the total size of the output is unknown in advance.

Option 1 is efficient and light-weigth, but it would possibly exclude too many applications using FORTRAN < 2003.
Option 2 is laborious with all the book keeping.

Do you favor one of the above options or do see another way?
Please, let me know your thoughts.

File: src/testdrive.F90 (hunk 7 above)
Comment (QGhd) [reply src/testdrive.F90:551]: My suggestion is that calling `junitxml_open_file()` switches a global variable from `.false.` to `.true`. If `.true` the output file is written. This requires that any of the `junitxml_write_*()` routines needs to respect this flag.
Let me know, if you propose a different way.
If not, I will implement as suggested.
Please, let me know your thoughts.

File: src/testdrive.F90 (hunk 9 above)
Comment (QGhd) [reply src/testdrive.F90:561]: I am not sure about the dynamic size of the string variable. Please, refer to my response here: https://github.com/fortran-lang/test-drive/pull/27#discussion_r1315170193.
Please, let me know your thoughts.

File: src/testdrive.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (QGhd) [reply src/testdrive.F90]: Please, let me know your thoughts.

File: src/JUnit.xml
Comment (QGhd) [reply src/JUnit.xml:1]: Please, let me know your thoughts.

File: src/JUnit.xsd
Comment (QGhd) [reply src/JUnit.xsd:1]: Please, let me know your thoughts.

File: src/jenkins-junit.xsd
Comment (QGhd) [reply src/jenkins-junit.xsd:1]: Please, let me know your thoughts.

**Comment by QGhd** [comment]
@awvwgk: With respect to your change request:
> Thanks, this looks pretty straight forward to add and I would be happy to accept this in test-drive. One general comment, the generation of the JUnit.xml is incremental, which means if the test run is aborted the partly generated JUnit.xml file is always invalid. Could we have a process which guarantees that the JUnit.xml file can only be written completely?

A delayed output of the file at the end of application is fmho not a reasonable option (see https://github.com/fortran-lang/test-drive/pull/27#discussion_r1315170193). I can try to delete the file, if it could not be written completely in case of an error.
Please, let me know your thoughts.

File: src/testdrive.F90 (hunk 6 above)
Comment (jvdp1) [reply src/testdrive.F90:568]: My suggestion is to use the Option 1 (i.e. `character(len=:), allocatable` variable). Applications written in older Fortran could still use this library, provided that the used compiler support Fortran 2003+.

File: src/testdrive.F90 (hunk 7 above)
Comment (jvdp1) [reply src/testdrive.F90:551]: That is an option.
Another possible option is to store the content in an `optional` `character` variable, and to write this variable at the end of all tests.
The `optional` variable would allow backward compatibility.

E.g.,
```fortran
character(len=:), allocatable :: xml_
        write(error_unit, fmt) "Suite:", testsuites(is)%name
        call run_selected(testsuites(is)%collect, test_name, error_unit, stat, xml =  xml_)
        call junitxml_write_testsuite(xml_)
```

**Comment by jvdp1** [comment]
> @awvwgk: With respect to your change request:
> 
> > Thanks, this looks pretty straight forward to add and I would be happy to accept this in test-drive. One general comment, the generation of the JUnit.xml is incremental, which means if the test run is aborted the partly generated JUnit.xml file is always invalid. Could we have a process which guarantees that the JUnit.xml file can only be written completely?
> 
> A delayed output of the file at the end of application is fmho not a reasonable option (see [#27 (comment)](https://github.com/fortran-lang/test-drive/pull/27#discussion_r1315170193)). I can try to delete the file, if it could not be written completely in case of an error. Please, let me know your thoughts.

My proposition to store the XML output in an `optional` variable and to write it to a file after running all tests could solve this issue.
